File: 22 Subarray Sum Equals K/Subarray Sum Equals K.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function (nums, k) {
    let cumSum = 0;
    let mp = new Map();
    let cnt = 0;
    for (let i = 0; i < nums.length; i++) {
        cumSum += nums[i];
        if (cumSum == k)
            cnt++;
        if (mp.has(cumSum - k))
            cnt += mp.get(cumSum - k);
        let x = 0;
        if (mp.has(cumSum))
            x = mp.get(cumSum);
        mp.set(cumSum, x + 1);
    }
    return cnt;
};

File: 22 Subarray Sum Equals K/Subarray Sum Equals K.cpp
class Solution
{
public:
    int subarraySum(vector<int> &nums, int k)
    {
        int cumSum = 0;
        map<int, int> mp;
        int cnt = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            cumSum += nums[i];
            if (cumSum == k)
                cnt++;
            if (mp.find(cumSum - k) != mp.end())
                cnt += mp[cumSum - k];
            mp[cumSum]++;
        }
        return cnt;
    }
};